library/core/src/num/dec2flt/decimal.rs

```diff
@@ -0,0 +1,384 @@
+    pub negative: bool,
+    /// If the number of significant digits stored in the decimal is truncated.
+    pub truncated: bool,
+    /// Buffer of the raw digits, in the range [0, 9].
+    pub digits: [u8; Self::MAX_DIGITS],
+}
+
+impl Debug for Decimal {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Decimal")
+            .field("num_digits", &self.num_digits)
+            .field("decimal_point", &self.decimal_point)
+            .field("negative", &self.negative)
+            .field("truncated", &self.truncated)
+            .field("digits", &(&self.digits[..self.num_digits]))
+            .finish()
+    }
+}
+
+impl PartialEq for Decimal {
+    fn eq(&self, other: &Self) -> bool {
+        self.num_digits == other.num_digits
+            && self.decimal_point == other.decimal_point
+            && self.negative == other.negative
+            && self.truncated == other.truncated
+            && self.digits[..] == other.digits[..]
+    }
+}
+
+impl Eq for Decimal {}
+
+impl Default for Decimal {
+    fn default() -> Self {
+        Self {
+            num_digits: 0,
+            decimal_point: 0,
+            negative: false,
+            truncated: false,
+            digits: [0; Self::MAX_DIGITS],
+        }
+    }
+}
+
+impl Decimal {
+    /// The maximum number of digits required to unambiguously round a float.
+    ///
+    /// For a double-precision IEEE-754 float, this required 767 digits,
+    /// so we store the max digits + 1.
+    ///
+    /// We can exactly represent a float in radix `b` from radix 2 if
+    /// `b` is divisible by 2. This function calculates the exact number of
+    /// digits required to exactly represent that float.
+    ///
+    /// According to the "Handbook of Floating Point Arithmetic",
+    /// for IEEE754, with emin being the min exponent, p2 being the
+    /// precision, and b being the radix, the number of digits follows as:
+    ///
+    /// `−emin + p2 + ⌊(emin + 1) log(2, b) − log(1 − 2^(−p2), b)⌋`
+    ///
+    /// For f32, this follows as:
+    ///     emin = -126
+    ///     p2 = 24
+    ///
+    /// For f64, this follows as:
+    ///     emin = -1022
+    ///     p2 = 53
+    ///
+    /// In Python:
+    ///     `-emin + p2 + math.floor((emin+ 1)*math.log(2, b)-math.log(1-2**(-p2), b))`
+    pub const MAX_DIGITS: usize = 768;
+    /// The max digits that can be exactly represented in a 64-bit integer.
+    pub const MAX_DIGITS_WITHOUT_OVERFLOW: usize = 19;
+    pub const DECIMAL_POINT_RANGE: i32 = 2047;
+
+    /// Append a digit to the buffer.
+    pub fn try_add_digit(&mut self, digit: u8) {
+        if self.num_digits < Self::MAX_DIGITS {
+            self.digits[self.num_digits] = digit;
+        }
+        self.num_digits += 1;
+    }
+
+    /// Trim trailing zeros from the buffer.
+    pub fn trim(&mut self) {
+        while self.num_digits != 0 && self.digits[self.num_digits - 1] == 0 {
+            self.num_digits -= 1;
+        }
```

Comment:
This can panic. Imagine when `num_digits == MAX_DIGITS`. You'd call `decimal.try_add_digit(0)`, which would correctly not add it to `digits` (because it's full), but still increment `num_digits`. Then, call `trim()`, which sees that num_digits != 0 and promptly panics when you index into digits.
```suggestion
    pub fn try_add_digit(&mut self, digit: u8) {
        if self.num_digits < Self::MAX_DIGITS {
            self.digits[self.num_digits] = digit;
            self.num_digits += 1;
        }
    }

    /// Trim trailing zeros from the buffer.
    pub fn trim(&mut self) {
        while self.num_digits != 0 && self.digits[self.num_digits - 1] == 0 {
            self.num_digits -= 1;
        }
```

Reply:
Great catch, I should probably file an issue in fast-float-rust as well on this...

Reply:
Ok this generates more efficient assembly anyway, so there's no reason not to implement this change.

**New**

```asm
example::Decimal::try_add_digit:
        mov     rax, qword ptr [rdi]
        cmp     rax, 767
        ja      .LBB0_2
        mov     byte ptr [rdi + rax + 14], sil
        mov     rax, qword ptr [rdi]
.LBB0_2:
        add     rax, 1
        mov     qword ptr [rdi], rax
        ret
```

**Original**

```asm
example::Decimal::try_add_digit:
        mov     rax, qword ptr [rdi]
        cmp     rax, 767
        ja      .LBB0_2
        mov     byte ptr [rdi + rax + 14], sil
        add     qword ptr [rdi], 1
.LBB0_2:
        ret
```

Reply:
Actually, this has different logic. We can't do `self.num_digits <= Self::MAX_DIGITS`, because that would cause an indexing error in the cause where `self.num_digits == Self::MAX_DIGITS`, and we can't do `self.num_digits < Self::MAX_DIGITS`, since that would prevent incrementing to `MAX_DIGITS`, which then causes significant issues elsewhere.

A quick test on values with a lot of digits will demonstrate this:

```rust
let value = dec2flt::<f32>(format!("{}e-38", append_zeros(f1, 656))).unwrap();
assert_eq(value.to_bits(), 0x7ffffe);   // Fails, actual value is 0xccccd
```